### PR TITLE
update for KUBE-IPTABLES-HINT (and other 2022-ness)

### PR DESCRIPTION
**Pull request by danwinship**
[KEP-3178](https://github.com/kubernetes/enhancements/issues/3178) will (eventually) make kubelet create fewer iptables rules so we need to update iptables-wrappers to cope with that.

Also, misc updates for the last few years of iptables evolution...

/assign @dcbw

### CODE REVIEW

File: test/Dockerfile.test-debian [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 FROM debian:buster
 
-ARG INSTALL_ARGS
+ARG INSTALL_ARGS=
```
Comment (aojea) [inline test/Dockerfile.test-debian:19]: is this `=`  right?
I've never seen this format, but doesn't mean isnot right

File: iptables-wrapper-installer.sh [hunk 2]
```diff
@@ -99,16 +99,34 @@ cat > "${sbin}/iptables-wrapper" <<EOF
...
+# iptables-nft, because we can check that more efficiently and
+# it's more common these days.
+nft_kubelet_rules=\$( (iptables-nft-save -t mangle || true; ip6tables-nft-save -t mangle || true) 2>/dev/null | grep -E '^:(KUBE-IPTABLES-HINT|KUBE-KUBELET-CANARY)' | wc -l)
+if [ "\${nft_kubelet_rules}" -ne 0 ]; then
     mode=nft
+else
+    # Check for kubernetes 1.17-or-later with iptables-legacy. We
+    # can't pass "-t mangle" to iptables-legacy-save because it would
+    # cause the kernel to create that table if it didn't already
+    # exist, which we don't want. So we have to grab all the rules
```
Comment (aojea) [inline iptables-wrapper-installer.sh:114]: :upside_down_face:

File: test/Dockerfile.test-debian (hunk 1 above)
Comment (danwinship) [reply test/Dockerfile.test-debian:19]: Yeah, it's just setting the default value to "". Which is what it was before, but podman complains now if you declare an ARG without giving it an explicit default value and then also don't set it from the command line:

```
danw@p1:iptables-wrappers (modernize)> make check-debian
./test/run-test.sh --build-fail debian
    docker build --no-cache -q -t iptables-wrapper-test-debian -f test/Dockerfile.test-debian .
    time="2022-04-11T09:12:43-04:00" level=warning msg="missing \"INSTALL_ARGS\" build argument. Try adding \"--build-arg INSTALL_ARGS=<VALUE>\" to the command line"
```

File: iptables-wrapper-installer.sh [hunk 3]
```diff
@@ -104,42 +99,37 @@ cat > "${sbin}/iptables-wrapper" <<EOF
...
+    legacy_kubelet_rules=\$( (iptables-legacy-save || true; ip6tables-legacy-save || true) 2>/dev/null | grep -E '^:(KUBE-IPTABLES-HINT|KUBE-KUBELET-CANARY)' | wc -l)
+    if [ "\${legacy_kubelet_rules}" -ne 0 ]; then
         mode=legacy
     else
-        mode=nft
+        # With older kubernetes releases there may not be any _specific_
```
Comment (squeed) [inline iptables-wrapper-installer.sh:119]: I would say it's safe to assume `KUBE-MARK-DROP` will always exist -- it was added in, AFAICT, v1.4.

That said, it probably makes sense to just fall back to "literally any rules exist at all".

**Comment by squeed** [comment]
looks good to me. Thanks for thinking to maintain this.

File: iptables-wrapper-installer.sh (hunk 3 above)
Comment (danwinship) [reply iptables-wrapper-installer.sh:119]: `KUBE-MARK-DROP` is going away :slightly_smiling_face: 
https://github.com/kubernetes/enhancements/pull/3179

**Pull request merged**